### ISSUE TITLE
Implement uploading of multpart/form-data

### ISSUE THREAD

**Issue by ptiz** [Moya/Moya#114]

**Comment by ashfurrow** [comment]
Great idea! I think Alamofire supports this already, so there is an opportunity to leverage their code while being clever about how we expose the functionality.

**Comment by ptiz** [comment]
I am afraid it is not, out of the box. But it would be great if there was something like Alamofire.upload at least.

**Comment by ashfurrow** [comment]
OK, cool. We can start there then :)

**Comment by aschuch** [comment]
There is [Alamofire support](https://github.com/Alamofire/Alamofire#uploading) for uploading and downloading files, so this would be straight forward I guess.

I anyone already working on this? Might need it in the following weeks so I might as well hack on it if you want.

**Comment by ashfurrow** [comment]
Help on this would be very much appreciated. I'm catching up on some work for the remainder of the week, and won't have time to look at it for a while.

**Comment by zdavison** [comment]
Also looking for this fairly soon. @aschuch do you have any progress on this? If not I can take a look.

**Comment by aschuch** [comment]
@zdavison I did not start on it yet. I probably don't get time to do it anytime soon though... :disappointed:

**Comment by zdavison** [comment]
@aschuch that's ok, i'll take a look here, we need it in the next week or so.

**Comment by ashfurrow** [comment]
This is a fundamentally new behaviour for Moya. I'm trying to figure out the best way to add it to our existing API without overly complicating things. 

If we want to use the existing `request()` function, then we would need to have a way to distinguish targets between upload and regular HTTP requests. We _could_ add `.Upload` to the `Moya.Method` enum, which would be easy, but would break parity with the HTTP verbs and the `Alamofire.Method` enum. 

On the other hand, if we add a new optional property to `MoyaTarget` for uploading (say, we call it `uploadData`), we could ignore the method of the target and assume it's uploading based on the presence of a non-`nil` `uploadData`. 

In either case, we would still need to figure out the progress callback. Maybe we could incorporate it into the cancellable token somehow maybe? 

Neither option really sits with me. We could, instead, have a new `upload` method on the `MoyaProvider` class. We would still need the optional `uploadData` property, the API would become _slightly_ more complicated, and it would be some duplicated code in `MoyaProvider` and its subclasses. 

Anyone have any feels about this?

**Comment by alcarvalho** [comment]
If only Swift had support for default parameters to Enums, we could have `.POST` and others with a optional parameter with a default to `nil`. 

But even if we don't have that support, we could just add more verbs that accept that to `Moya.Method`.

**Comment by zdavison** [comment]
FWIW we ended up just doing it by hand (since we only need it for one request) and having a special case that returns a MoyaResponse until Moya/Alamofire support it natively.

**Comment by SpacyRicochet** [comment]
@zdavison Can we see what your solution looks like? I tried hacking it in an endpointResolver, but I hit a wall when I figured out the `Endpoint` doesn't seem to know about any of the parameters it's been given in by the `enum`. Now, creating a custom, plain NSURLRequest seems the best option. 

But you returning a `MoyaResponse` sounds intriguing, but I don't get what that would look like.

@ashfurrow 

> On the other hand, if we add a new optional property to MoyaTarget for uploading (say, we call it uploadData), we could ignore the method of the target and assume it's uploading based on the presence of a non-nil uploadData.

I like this one. Beyond the fact that we still need a call-back of some kind, it fits very well within the existing way of creating a `Target`.

> Neither option really sits with me. We could, instead, have a new upload method on the MoyaProvider class. We would still need the optional uploadData property, the API would become slightly more complicated, and it would be some duplicated code in MoyaProvider and its subclasses.

This also sounds good, though slightly confused on how this would interact with a `Target`. Do suddenly all `Target` cases have the ability to upload data if you call them with this method? Should the method check if the request is a .POST or not?

**Comment by zdavison** [comment]
```
  // Multipart image uploading not supported by Moya/Alamofire, temporary stand in
  internal static func ns_multipartImageUpload(urlString: String, image: UIImage, imageKind: ImageKind, progressHandler: ((Float) -> Void)?) -> RACSignal{
    
    let imageData       = UIImageJPEGRepresentation(image, 9.5)
    
    return RACSignal.createSignal{
      (subscriber) in
      
      Just.post(
        urlString,
        headers:    ["X-Auth-Token":  DropLogic.sharedInstance.apiToken!],
        data:       ["sizes":         imageKind.rawValue],
        files:      ["file":          HTTPFile.Data("image.jpeg", imageData, nil)],
        asyncProgressHandler: {
          (progress) in
          progressHandler?(progress.percent)
        },
        asyncCompletionHandler: {
          (result) in
          
          if let error = result.error {
            subscriber.sendError(error)
            return
          }
          
          let moyaResponse = MoyaResponse(statusCode: result.statusCode!, data: result.content!, response: result.response)
          subscriber.sendNext(moyaResponse)
          subscriber.sendCompleted()
        }
      )
      
      return RACDisposable()
    }
  }
```

Using [Just](https://github.com/JustHTTP/Just.git) and [ReactiveCocoa](https://github.com/ReactiveCocoa/ReactiveCocoa).

**Comment by SpacyRicochet** [comment]
Didn't know about Just :) That's going to be useful. Thanks!

**Comment by clooth** [comment]
Has there been any work on this yet? :pray:

**Comment by aschuch** [comment]
Not yet, do you want to give it a shot and open a PR @clooth?

**Comment by clooth** [comment]
I'll give it a try this weekend, need to look around the source a bit so it feels enough Moya-like.

**Comment by mgroh** [comment]
BTW Alamofire now [supports uploading multipart/form-data] natively.(https://github.com/Alamofire/Alamofire#uploading-multipartformdata.)

**Comment by GuidoHendriks** [comment]
I would really like this to be part of Moya. Might give it as shot, but I think I need a better understanding of the project before that will work... Is anyone else making progress on this?

**Comment by ashfurrow** [comment]
Cool, that would be great! Feel free to ask questions in issues.

As for an implementation of this, I figure there's two part "entry points" where you could specify multipart upload: the `MoyaTarget` protocol or the `request` function. I think I'd prefer adding another function, or expanding the existing `request` function to handle it. Not sure of specifics, but you get the idea.

**Comment by swizzlr** [comment]
Let's make this happen!

How about creating a subprotocol, `MultipartFormUploadable` (I think we should get rid of this Moya namespacing too, but that's kinda just me) that inherits from `MoyaTarget` – we overload `request` to take the more specific type, and to use it, and require an extra function to vend either an NSData, NSStream or a closure that returns those things to supply the body.

This might be the groundwork for a more strongly typed set of targets that more closely specify their methods of encoding and requesting, though that is perhaps an exercise best left to the reader.

**Comment by JRHeaton** [comment]
Played around with your ideas a bit, @swizzlr. A problem I'm running into is that if using something like:
```swift
public protocol MultipartFormUploadable: MoyaTarget { }

public enum UploadDataProvider {
    case Data(NSData)
    case File(NSURL)
    case Stream(NSInputStream)
}

extension MoyaProvider where Target: MultipartFormUploadable {
    public func request(target: Target, provider: UploadDataProvider, completion: Moya.Completion) -> Cancellable
}
```
Then your _entire_ target type is then needing to conform to this protocol, or you need a second provider and target type for your multipart endpoints. While it doesn't seem like a necessarily common occurrence with a buffer of common sense, someone could technically call this version of `request()` with a `POST` target that is **not** multipart with a random (valid) data provider, since there's no way to distinguish between multipart and non-multipart enum target cases. This also means that the "original" `request()` can be called with a multipart case, which is also wrong. Any better ideas for separation and compile-time verification of these types of endpoint targets?

**Comment by ashfurrow** [comment]
Yeah, this is why I think adding another `request` function specific to multipart uploading would be the best approach.

**Comment by JRHeaton** [comment]
@ashfurrow Yeah, that is what I demonstrated above in the extension, but saw potential confusion with. Since the entire enum type must conform to `MultipartFormUploadable`, there's no way to determine whether a specific enum case should be able to be used for a multipart upload or not. This means it's not "perfectly strict" in that it can't warn or error if you try to misuse it in this way (misuse meaning call the overloaded version of `request()` in the extension with a data provider). If you do so, you're stuck waiting potentially for the server to WTF at you for trying to upload data to a POST endpoint that doesn't take it, if that will even happen. In short, there's something that we know would at compile time, but can't enforce with the compiler in any way that I can think of, at least with targets as enum types.

**Comment by GuidoHendriks** [comment]
Maybe we should not focus on `multipart/form-data`, but on uploads in general. In Alamofire there are [4 types of uploads](https://github.com/Alamofire/Alamofire#uploading):

* File
* Data
* Stream
* MultipartFormData

I would say we add some kind of request type, which can be either `Request` or `Upload` (we could add  `Download` too). Based on the request type it would execute the right Alamofire methods with the needed data from the target. That would also mean adding some kind of `UploadType` enum for all the supported upload types.

This way you maintain the abstraction. If you were adding a request specifically for uploads, you'd be picking the request type when executing a request, but that's something that the target should know.

Please do correct me if I'm wrong, but this seems like a step in the right direction to me.

**Comment by colinta** [comment]
This makes total sense to me!

**Comment by petester42** [comment]
Makes sense to add the type of request to the target.

**Comment by clooth** [comment]
Yeah, I'm sorry to say that I don't have time for this after all.

**Comment by colinta** [comment]
David, could you attach an example of the Ello.Target that would need to use Request and Upload types? I want to be sure I'm following you...

**Comment by drw86312** [comment]
@colinta pretty much any endpoints that could take optional data. I may not be completely understanding @GuidoHendriks solution, but I'd like to not have to write multiple enum cases for, say a "Create Account" request that may or may not take image data.  Essentially, I want one "Create Account" endpoint to which I can optionally pass one or multiple NSData, NSURL, NSInputStream parameters.  

Essentially, I'd like to avoid extraneous enum cases like this...
```swift
public enum API: MoyaTarget {

    case CreateAccount(name: String, type: String, description: String)

    case CreateAccountWithImageData(name: String, type: String, description: String, imageData : NSData)
}

extension API  {

    public var requestType: RequestType {
        switch self {
        case .CreateAccount:
            return .Request

        case .CreateAccountWithImageData:
            return .Upload
        }
}
```

**Comment by colinta** [comment]
Yup that's about what I was imagining, and I wouldn't want those two endpoints either.

**Comment by GuidoHendriks** [comment]
@drw86312 Why would you need an extra endpoint for that? You could instead just use an optional:

```swift
public enum API: MoyaTarget
{
    case CreateAccount(name: String, imageData: NSData?)

    public var requestType: RequestType {
        switch self {
        case .CreateAccount(_, let imageData) where imageData != nil:
            return .Upload
        default:
            return .Request
        }
    }
}

API.CreateAccount(name: "Guido", imageData: nil).requestType      // .Request
API.CreateAccount(name: "Guido", imageData: NSData()).requestType // .Upload
```

**Comment by GuidoHendriks** [comment]
@zdavison Would you mind sharing your solution on manual implementation?

**Comment by Alexey-Matjuk** [comment]
Looking forward use this feature some day. As temporary solution I:
1) added file property in MoyaTarget protocol:
```
public protocol MoyaTarget {
     ...
     var file: NSURL? { get }
}
```
2) use result of [1] to determine type of request:
```
    func sendRequest(target: Target, request _request: NSURLRequest, completion: Moya.Completion) -> CancellableToken {
        let request: Request
        if let file = target.file {
            request = manager.upload(_request, file: file)
        } else {
            request = manager.request(_request)
        }
        ...
```

**Comment by mMaxy** [comment]
Hey, guys. Any progress on this issue? Is there any solution that is working and can be used?
Anyway, upvoting!

And thanks for so cool artefact, Moya literally saved me days of work and tons of headache!

**Comment by startupthekid** [comment]
Getting in way late to the party here but I second @GuidoHendriks' idea of just adding a `RequestType` enum with two cases `.Upload` and `.Request` and then a second enum, `UploadType` with four cases: `.File`, `.Data`, `.Stream`, `.Multipart`.

We could then add two variables to `TargetType`:

```
public var requestType: RequestType { get }
public var uploadType: UploadType? 
```
 where `uploadType` would only be checked if `requestType` returns `.Upload`. What I like a lot about this approach is it's easy to implement, it keeps the abstraction, and makes migrations much easier since all people now have to do is add the two new variables to conform. 

My only concern is as mentioned, how to show progress of a upload request.

**Comment by aschuch** [comment]
@startupthekid I like that.

> where uploadType would only be checked if requestType returns .Upload. What I like a lot about this approach is it's easy to implement, it keeps the abstraction, and makes migrations much easier since all people now have to do is add the two new variables to conform.

We could even go as far as adding a default implementation with sensible defaults to the added protocol vars, which would "upgrade" any existing implementations.

**Comment by startupthekid** [comment]
Exactly. Anything that doesn't explicitly say they're an upload is just a request. People can opt into the behavior as needed.

**Comment by justinmakaila** [comment]
I'm thinking we should `TargetType` generally the same and steal some patterns from `Alamofire`...

If we add `ParameterEncoding` to `TargetType`, then we could describe a multipart like so:

```
public enum Target {
  ...
  /// Represents a data part upload.
  /// Requires a data part to upload, and the name of the part
  case UploadData(NSData, String, String, String)
}

extension Target: TargetType {
  var parameterEncoding: ParameterEncoding {
    switch self {
    case .UploadPhoto(let data, let partName, let fileName, let mimeType):
      return .Multipart({ formData in
        formData.appendBodyPart(data: data, name: partName, fileName: fileName, mimeType)
      })
    default:
      return .JSON
    }
  }
}
```
I'm going to spend a little time poking around the internals to see what makes the most sense. Personally, this is my preferred API use case. I think it makes a bit more sense than introducing new abstractions that are ignored by 90% implementations, and also gives the developer a bit more power in handling the encoding on a per-request basis.

**Comment by ashfurrow** [comment]
Fixed in #483.
